File: imap-reader/src/main/kotlin/de/ideasinlogic/tools/imappdf/OCRWrapper.java
package de.ideasinlogic.tools.imappdf;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.io.File;
import java.io.IOException;
import java.util.Optional;

/**
 * Wrapper for the OCR process.
 */
public class OCRWrapper {
  private static Logger log = LoggerFactory.getLogger(OCRWrapper.class);

  /**
   * Run the OCR.
   *
   * @param input an input PDF.
   * @param lang  language to use.
   * @return an PDF file with text-annotations, if successful.
   */
  Optional<File> runOCR(File input, String lang) {
    try {
      File output = File.createTempFile("scan-", ".pdf");
      ProcessBuilder pb = new ProcessBuilder(
          "runocr.sh",
          // "--output-type", "pdf",
          // "--pdf-renderer", "tesseract",
          lang,
          input.getAbsolutePath(),
          output.getAbsolutePath());
      pb.inheritIO();
      Process process = pb.start();
      log.debug("Started OCR process");
      int waitFor = process.waitFor();
      if (waitFor != 0) {
        log.warn("failed to process " + waitFor);
        //noinspection ResultOfMethodCallIgnored
        output.delete();
        return Optional.empty();
      }
      log.debug("finished process " + waitFor + " and created file " + output.getAbsolutePath());
      return Optional.of(output);
    } catch (InterruptedException | IOException e) {
      log.debug("failure during processing " + input.getAbsolutePath());
      return Optional.empty();
    }
  }

}

File: imap-reader/src/main/kotlin/de/ideasinlogic/tools/imappdf/OCRWrapper.kt
package de.ideasinlogic.tools.imappdf

import mu.KotlinLogging
import java.io.File
import java.io.IOException

/**
 * Wrapper for the OCR process.
 */
class OCRWrapper {
	/**
	 * Run the OCR.
	 *
	 * @param input an input PDF.
	 * @param lang  language to use.
	 * @return an PDF file with text-annotations, if successful.
	 */
	fun runOCR(input: File, lang: String?): File? {
		return try {
			val output = File.createTempFile("scan-", ".pdf")
			val pb = ProcessBuilder(
				"runocr.sh",
				lang ?: "deu",
				input.absolutePath,
				output.absolutePath
			)
			pb.inheritIO()
			val process = pb.start()
			log.debug { "Started OCR process" }
			val waitFor = process.waitFor()
			if (waitFor != 0) {
				log.warn { "failed to process $waitFor" }
				output.delete()
				return null
			}
			log.debug { "finished process " + waitFor + " and created file " + output.absolutePath }
			output
		} catch (e: InterruptedException) {
			log.debug { "failure during processing " + input.absolutePath }
			return null
		} catch (e: IOException) {
			log.debug { "failure during processing " + input.absolutePath }
			return null
		}
	}

	companion object {
		private val log = KotlinLogging.logger { }
	}
}